### PR TITLE
Add executor field to the DB and parameter to the operators

### PR DESCRIPTION
**Pull request by o-nikolas**
This field will be used by AIP-61 code to store and retrieve which executor a task is meant to run on.

This PR adds the new field and parameter but does not make any use of it just yet. No documentation is provided, intentionally, since the AIP-61 feature is not yet complete. I am slowly implementing pieces instead of one giant PR. A warning is shown if a user attempts to provide a value for this parameter to any operators.

I've done manual UAT with these changes (hello world dag, triggere dag, mapped task, task flow, etc) and all seems to be functioning.

I have some open question in the code which I've annotated with comments in the PR as well.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/migrations/versions/0137_2_9_0_add_new_executor_field_to_db.py
Comment (o-nikolas) [inline airflow/migrations/versions/0137_2_9_0_add_new_executor_field_to_db.py]: Looking for feedback on this, thanks!

File: airflow/models/abstractoperator.py
Comment (o-nikolas) [inline airflow/models/abstractoperator.py]: Looking for feedback on this, thanks!

File: airflow/models/baseoperator.py
Comment (o-nikolas) [inline airflow/models/baseoperator.py]: Looking for feedback on this, thanks!

File: airflow/models/taskinstance.py
Comment (o-nikolas) [inline airflow/models/taskinstance.py]: I believe this should be nullable, but just flagging this to be doublechecked in review.

File: airflow/models/baseoperator.py
Comment (o-nikolas) [inline airflow/models/baseoperator.py]: I'm just noticing now that others are using NOTSET, instead of None. I'll update this unless anyone else things otherwise.

Comment (uranusjr) [reply airflow/models/baseoperator.py]: This should use `NOTSET` because we want to fallback to `default_args` when the argument is not used by the user, instead of the default here. This is why we have the `_PARTIAL_DEFAULTS` dict above—the fallback order is

1. User-supplied value
2. The DAG’s `default_args`
3. Value from `_PARTIAL_DEFAULTS`

File: airflow/models/baseoperator.py
Comment (uranusjr) [reply airflow/models/baseoperator.py]: I said this before, the entire dict is pretty sketchy.

File: airflow/models/baseoperator.py
Comment (uranusjr) [inline airflow/models/baseoperator.py]:
```suggestion
                f"supported, the value {executor!r} will have no effect"
```

File: airflow/models/taskinstance.py
Comment (uranusjr) [reply airflow/models/taskinstance.py]: What would the NULL case mean? Compared to, say, an empty string, or the default executor value?

File: airflow/models/abstractoperator.py
Comment (syedahsn) [reply airflow/models/abstractoperator.py]: Why do we want this to be NULL?

File: tests/www/views/test_views_tasks.py
Comment (ferruzzi) [inline tests/www/views/test_views_tasks.py:1045]: Not expecting you to change it but wow is that test incredibly repetitive.  Maybe after this is merged I'll do a quick PR to move all the fixed values into a dict and just unpack then in each of these.  Would make it much easier to read and the only values that are called out explicitly would be the ones that are changing and worth paying specific attention to.

File: airflow/models/taskinstance.py
Comment (ferruzzi) [reply airflow/models/taskinstance.py]: I don't think nullable makes sense here, as uranusjr said..  can't think of a case where "no executor" would be valid?

File: airflow/models/baseoperator.py
Comment (ferruzzi) [reply airflow/models/baseoperator.py]: Yeah, that's awkward.     I think it goes in there, seems to make sense to me.

File: airflow/models/baseoperator.py
Comment (ferruzzi) [inline airflow/models/baseoperator.py]: Nittiest of nitpicks.  Some are capitalized, some aren't, feel free to ignore

```suggestion
    :param executor: Which executor to target when running this task. NOT YET SUPPORTED
```

**Review by ferruzzi — COMMENTED** [review]
Left a couple comments.  I've never done a db migration, so I can't review that part.

File: airflow/models/baseoperator.py
Comment (potiuk) [reply airflow/models/baseoperator.py]: If in doubt, mimic what you have there. _comps is used to calculate hash. if you do not add it here, then two identical tasks with different executor might have same hash. It's certainly wrong no matter what hash is used for. So yes it should be added here.

File: airflow/migrations/versions/0137_2_9_0_add_new_executor_field_to_db.py
Comment (potiuk) [reply airflow/migrations/versions/0137_2_9_0_add_new_executor_field_to_db.py]: Looking at other migrations - default is the Python default value assigned when the object is created  not a server side default - which means that existing rows will get the default. And there are already a number of migrations that do not have defaults, so if you do not want to set a value, omitting default should work. But of course testing it is a good idea.

File: airflow/models/taskinstance.py
Comment (potiuk) [reply airflow/models/taskinstance.py]: Yes. Executor should ratehr be not null and historical value shoudl be set to default executor at migration time.

File: airflow/migrations/versions/0137_2_9_0_add_new_executor_field_to_db.py
Comment (potiuk) [reply airflow/migrations/versions/0137_2_9_0_add_new_executor_field_to_db.py]: But I guess we want to set default in this case (see below) - so likely we need to set default to defailt executor config.

File: airflow/models/abstractoperator.py
Comment (o-nikolas) [reply airflow/models/abstractoperator.py]: Good question, it's covered in more detail in the AIP (I wish I could hyper link to one specific section or comment...) but essentially NULL, means that no specific executor was requested for the task. In those cases we fall back to the historical/default behaviour of Airflow, which is to use whatever the currently configured environment level executor is, which may change at any time (if the user decides to switch it).

File: airflow/models/baseoperator.py
Comment (o-nikolas) [reply airflow/models/baseoperator.py]: > If in doubt, mimic what you have there. _comps is used to calculate hash. if you do not add it here, then two identical tasks with different executor might have same hash. It's certainly wrong no matter what hash is used for. So yes it should be added here.

Yupp, this was my understanding as well. Thanks for the verification!

File: airflow/models/taskinstance.py
Comment (o-nikolas) [reply airflow/models/taskinstance.py]: This is actually a nuanced bit that's easy to get confused by. The default must be NULL/empty to maintain the behaviour we have today in the base case. I covered this in the AIP and I think that is the best language I've come up with, so I've attached that below. As well as comment threads from Jens and Jarek who +1 the idea during review:

Explanation from AIP:
Screenshot from 2024-03-14 11-21-29

Comment thread:
Screenshot from 2024-03-14 11-21-42
Screenshot from 2024-03-14 11-21-51
Screenshot from 2024-03-14 11-22-00

File: airflow/models/baseoperator.py
Comment (o-nikolas) [reply airflow/models/baseoperator.py]: Thanks for the details, I will indeed move this to use NOTSET :+1:

File: airflow/models/taskinstance.py
Comment (o-nikolas) [reply airflow/models/taskinstance.py]: @potiuk Does the above make sense after I pulled in more context?

Comment (potiuk) [reply airflow/models/taskinstance.py]: Ah right. That makes sense (again)

**Review by syedahsn — APPROVED** [review]

**Comment by o-nikolas** [comment]
@uranusjr @potiuk I've addressed all the comments, done a fresh rebase and tests are passing. Can I get another a approval to merge this one?

**Review by potiuk — APPROVED** [review]

**Review by eladkal — APPROVED** [review]

**Comment by eladkal** [comment]
@o-nikolas can you rebase?

**Comment by potiuk** [comment]
Yep. We cannot merge it without resolving conflict :(

**Comment by o-nikolas** [comment]
Yupp, will do, it was all green and no conflicts when I left on Friday :sob: I'll do the rebase today

**Comment by potiuk** [comment]
BTW. We discussed in #release-management - that probably we want to hold-on with merging it as we are planning to move the v2-9-test to start from latest `main` rather than cherry-pick changes for 2.9.0b2 - so this one could be merged after we move the branch (cc: @ephraimbuddy ) - as this one adds migration, it makes sense to do in for 2.10 as this is where it will be actually useful.

**Comment by o-nikolas** [comment]
> BTW. We discussed in #release-management - that probably we want to hold-on with merging it as we are planning to move the v2-9-test to start from latest `main` rather than cherry-pick changes for 2.9.0b2 - so this one could be merged after we move the branch (cc: @ephraimbuddy ) - as this one adds migration, it makes sense to do in for 2.10 as this is where it will be actually useful.

Okay, that sounds reasonable, when is that branching going to happen? It's helpful to have this in main soon-ish so that I can create more dev branches off of it rather than have several layers of branches based on branches (which is possible of course, but just kind of a pain).

**Comment by ephraimbuddy** [comment]
> > BTW. We discussed in #release-management - that probably we want to hold-on with merging it as we are planning to move the v2-9-test to start from latest `main` rather than cherry-pick changes for 2.9.0b2 - so this one could be merged after we move the branch (cc: @ephraimbuddy ) - as this one adds migration, it makes sense to do in for 2.10 as this is where it will be actually useful.
> 
> Okay, that sounds reasonable, when is that branching going to happen? It's helpful to have this in main soon-ish so that I can create more dev branches off of it rather than have several layers of branches based on branches (which is possible of course, but just kind of a pain).

We are waiting for https://github.com/apache/airflow/pull/38446 and one other PR. So probably later today

**Comment by o-nikolas** [comment]
Okay @potiuk, @ephraimbuddy 

The PR is rebased and green again. Can you merge it when the coast is clear re: 2.9.0 release planning? Thanks!

**Comment by potiuk** [comment]
Ups. We were supposed to merge it after we move branch :)

**Comment by ferruzzi** [comment]
I'm sorry! Want me to revert the commit?

**Comment by potiuk** [comment]
Yeah I think it's best now. 

BTW. I had no idea the button is there :)

**Comment by ferruzzi** [comment]
I have reverted my accidental merge here.   I don't see a way to reopen this PR so we may need to apply it again later by reverting the revert, unless  @o-nikolas opens a new PR?

**Comment by potiuk** [comment]
yep. @o-nikolas to open a PR or you can do it by cherry-picking the reverted commit

**Comment by ferruzzi** [comment]
I have access to the fork/branch he submitted it from, [aws-mwaa:onikolas/aip-61/db_migration*](https://github.com/aws-mwaa/upstream-to-airflow/tree/onikolas/aip-61/db_migration), so I resubmitted it [here](https://github.com/apache/airflow/pull/38474) and verified that it shows his credit on the commits.

* I got that origin fork/branch from here:

image

**Comment by ferruzzi** [comment]
> BTW. I had no idea the button is there :)

I had never used it and was rather nervous LOL

**Pull request merged**
